[Opset15][Spec] Col2Im-15 specification (#23947)
openvinotoolkit/openvino: docs/articles_en/documentation/openvino-ir-format/operation-sets/operation-specs/movement/col2im-15.rst

* ``N`` is the batch dimension,
* ``C * Product(kernel_size)`` is the number of elements within a block (each block contains ``Product(kernel_size)`` vectors containing values from each channel ``C``),
* ``L`` is the total number of blocks calculated as follows:

.. math::

    L = \prod_{d=1}^{2} \left\lfloor \frac{({\text{{output\_size}}[d] + \text{{pads\_begin}}[d] + \text{{pads\_end}}[d] - \text{{dilation}}[d] \times (\text{{kernel\_size}}[d] - 1) - 1}}{{\text{{stride}}[d]}}) + 1\rfloor

where ``d`` is over all spatial dimensions.

The ``input`` blocks are being moved into the ``output`` tensor of shape ``(N, C, output_size[0], output_size[1])`` by combining the values contained in blocks.

**Attributes**:

* *output_size*

  * **Description**: controls the shape of the spatial dimensions of the output image.
  * **Range of values**: 1D tensor of non-negative integer numbers
  * **Type**: *T_IND*
  * **Required**: *yes*

* *kernel_size*

  * **Description**: size of the sliding blocks.
  * **Range of values**: 1D tensor of non-negative integer numbers
  * **Type**: *T_IND*

>   * **Required**: *yes*

Author: Done.

.. {#openvino_docs_ops_type_Col2Im_15}

Col2Im
===================

.. meta::
  :description: Learn about Col2Im-15 - data movement operation which combines sliding blocks into an image tensor.

**Versioned name**: *Col2Im-15*

**Category**: *Data movement*

**Short description**: *Col2Im* operation constructs an image based on ``input`` tensor containing sliding data blocks (blocks of the image) and desired ``output_size``. 

**Detailed description**

Consider an ``input`` tensor containing batches of image blocks of shape ``(N, C * Product(kernel_size), L)``, where:

* ``N`` is the batch dimension,
* ``C * Product(kernel_size)`` is the number of elements within a block (each block contains ``Product(kernel_size)`` vectors containing values from each channel ``C``),
* ``L`` is the total number of blocks calculated as follows:

.. math::

>     L = \prod_{d=1}^{2} \left\lfloor \frac{({\text{{output\_size}}[d] + \text{{pads\_begin}}[d] + \text{{pads\_end}}[d] - \text{{dilation}}[d] \times (\text{{kernel\_size}}[d] - 1) - 1}}{{\text{{stride}}[d]}}) + 1\rfloor

Author: ![image](https://github.com/openvinotoolkit/openvino/assets/55858107/0e996d3d-9251-4a58-9b3f-9d5b5ce46e54)

where ``d`` is over all spatial dimensions.

The ``input`` blocks are being moved into the ``output`` tensor of shape ``(N, C, output_size[0], output_size[1])`` by combining the values contained in blocks.

> Non-batched inputs are also supported, in which case the ``input`` has shape ``(C * Product(kernel_size), L)`` and the output has shape ``(C, output_size[0], output_size[1])``.

Author: I followed PyTorch's spec where `output_size` is always size of 2 (height and width). I can add an assert in Core implementation that its length is 2.

**Attributes**:

* *strides*

  * **Description**: stride in the sliding blocks in the input spatial dimensions.

>   * **Range of values**: 1D tensor of non-negative integer numbers

Author: Good catch, changed.

* *pads_begin*

  * **Description**: *pads_begin* is a number of zero-value pixels to add to the beginning along each axis. For example, *pads_begin* equal "1,2" means adding 1 pixel to the top of the input and 2 to the left of the input.
  * **Range of values**: 1D tensor of non-negative integer numbers
  * **Type**: *T_IND*
  * **Default value**: [0, 0]
  * **Required**: *no*

* *pads_end*

  * **Description**: *pads_end* is a number of zero-value pixels to add to the ending along each axis. For example, *pads_end* equal "1,2" means adding 1 pixel to the bottom of the input and 2 to the right of the input.
  * **Range of values**: 1D tensor of non-negative integer numbers
  * **Type**: *T_IND*
  * **Default value**: [0, 0]
  * **Required**: *no*

**Inputs**

* **1**: *data*

  * **Description**: A batched 3D tensor of type *T* and shape ``(N, C * Product(kernel_size), L)`` or an unbatched 2D tensor of type *T* and shape ``(C * Product(kernel_size), L)``. **Required.**
  * **Range of values**: 1D tensor of non-negative integer numbers
  * **Type**: *T*

* **2**: *output_size*

  * **Description**: controls the shape of the spatial dimensions of the output image. **Required.**

>   * **Range of values**: batched 3D or unbatched 2D tensor of non-negative integer numbers

Reviewer: It doesn't match to `shape_infer` in the other PR.

Reviewer 2: ```suggestion
  * **Range of values**: 1D tensor with 3 elements or 2 elements of non-negative integer numbers
```

Author: Corrected to `1D tensor of two non-negative integer numbers (height and width)`, following PyTorch's convention.

  * **Type**: *T_IND*

* **3**: *kernel_size*

  * **Description**: size of the sliding blocks. **Required.**
  * **Range of values**: 1D tensor of non-negative integer numbers
  * **Type**: *T_IND*

**Outputs**

* **1**: The output tensor the output image of type *T* and shape:

  * ``(N, C, output_size[0], output_size[1])`` in case of batched input,
  * ``(C, output_size[0], output_size[1])`` in case of non-batched input.

**Types**

* *T*: any supported data type.
* *T_IND*: ``int64`` or ``int32``.

**Examples**

All examples assume ``C = 3``.

*Example 1: default optional Parameters*

.. code-block:: xml
   :force:

    <layer ... type="Col2Im" ... >

>         <data output_size="16,16" kernel_size="2,2"/>

Reviewer 2: When `output_size` and `kernel_size` moved to the "Inputs" sections, then they should be provided as <input> in the IR examples.

.. code-block:: xml
   :force:

    <layer ... type="Col2Im" ... >
        <data output_size="16,16" kernel_size="3,3" dilation="2,2" pads_begin="1,1" pads_end="1,1" stride="2,2"/>
        <input>
            <port id="0" precision="I32">
                <dim>1</dim>     <!-- batch_axis -->
                <dim>27/dim>     <!-- C * Product(kernel_size) -->
                <dim>25</dim>    <!-- L -->
            </port>
        </input>
        <output>
            <port id="1" precision="I32">
                <dim>1</dim>     <!-- batch_axis -->
                <dim>3</dim>     <!-- C -->
                <dim>16</dim>    <!-- output_size[0] -->
                <dim>16</dim>    <!-- output_size[1] -->
            </port>
        </output>
    </layer>

*Example 3: non-default dilation and padding*

.. code-block:: xml
   :force:

    <layer ... type="Col2Im" ... >
        <data output_size="32,32" kernel_size="2,2" dilation="2,2" pads_begin="3,3" pads_end="3,3"/>
        <input>
            <port id="0" precision="I32">
                <dim>12</dim>    <!-- batch_axis -->
                <dim>12/dim>     <!-- C * Product(kernel_size) -->

>                 <dim>1296</dim>  <!-- L -->

Author: ```suggestion
                <dim>324</dim>  <!-- L -->
```

  * **Type**: *T_IND*
  * **Default value**: [0, 0]
  * **Required**: *no*

* *pads_end*

  * **Description**: *pads_end* is a number of zero-value pixels to add to the ending along each axis. For example, *pads_end* equal "1,2" means adding 1 pixel to the bottom of the input and 2 to the right of the input.
  * **Range of values**: 1D tensor of non-negative integer numbers
  * **Type**: *T_IND*
  * **Default value**: [0, 0]
  * **Required**: *no*

**Inputs**

* **1**: *data*

  * **Description**: A batched 3D tensor of type *T* and shape ``(N, C * Product(kernel_size), L)`` or an unbatched 2D tensor of type *T* and shape ``(C * Product(kernel_size), L)``. **Required.**
  * **Range of values**: 1D tensor of non-negative integer numbers
  * **Type**: *T*

* **2**: *output_size*

  * **Description**: controls the shape of the spatial dimensions of the output image. **Required.**
  * **Range of values**: 1D tensor of two positive integer numbers (height and width)
  * **Type**: *T_IND*

* **3**: *kernel_size*

  * **Description**: size of the sliding blocks. **Required.**
  * **Range of values**: 1D tensor of non-negative integer numbers

>   * **Type**: *T_IND*
> 
> **Outputs**
> 
> * **1**: The output tensor the output image of type *T* and shape:
> 
>   * ``(N, C, output_size[0], output_size[1])`` in case of batched input,
>   * ``(C, output_size[0], output_size[1])`` in case of non-batched input.
> 
> **Types**
> 
> * *T*: any supported data type.
> * *T_IND*: ``int64`` or ``int32``.

Reviewer 2: Detail:
As `T_IND` stands for "Type of index" and it is reused to describe also element types of the shapes, consider renaming it to `T_INT` (for integers).
```suggestion
  * **Type**: *T_INT*

**Outputs**

* **1**: The output tensor the output image of type *T* and shape:

  * ``(N, C, output_size[0], output_size[1])`` in case of batched input,
  * ``(C, output_size[0], output_size[1])`` in case of non-batched input.

**Types**

* *T*: any supported data type.
* *T_INT*: ``int64`` or ``int32``.
```

Author: Done.

            </port>
        </output>
    </layer>

*Example 2: non-default dilations, padding and strides*

For inputs ``output_size`` = [16, 16] and ``kernel_size`` = [3, 3]

.. code-block:: xml
   :force:

    <layer ... type="Col2Im" ... >
        <data dilations="2,2" pads_begin="1,1" pads_end="1,1" strides="2,2"/>
        <input>
            <port id="0" precision="I32">
                <dim>1</dim>     <!-- batch_axis -->
                <dim>27/dim>     <!-- C * Product(kernel_size) -->
                <dim>25</dim>    <!-- L -->
            </port>
            <port id="1" precision="I32">
                <dim>2</dim>     <!-- output_size -->
            </port>
            <port id="2" precision="I32">
                <dim>2</dim>     <!-- kernel_size -->
            </port>
        </input>
        <output>
            <port id="1" precision="I32">
                <dim>1</dim>     <!-- batch_axis -->
                <dim>3</dim>     <!-- C -->
                <dim>16</dim>    <!-- output_size[0] -->
                <dim>16</dim>    <!-- output_size[1] -->
            </port>
        </output>
    </layer>

*Example 3: non-default dilations and padding*

> For inputs ``output_size`` = [32, 32] and ``kernel_size`` = [2, 2]

Reviewer 2: For `Example 3` and `Example 4`, the ports for `output_size` and `kernel_size` inputs are still missing.

Author: I forgot to commit it :) It's fixed.

.. {#openvino_docs_ops_type_Col2Im_15}

Col2Im
===================

.. meta::
  :description: Learn about Col2Im-15 - data movement operation which combines sliding blocks into an image tensor.

**Versioned name**: *Col2Im-15*

**Category**: *Data movement*

**Short description**: *Col2Im* operation constructs an image based on ``input`` tensor containing sliding data blocks (blocks of the image) and desired ``output_size``. 

**Detailed description**

Consider an ``input`` tensor containing batches of image blocks of shape ``(N, C * Product(kernel_size), L)``, where:

* ``N`` is the batch dimension,
* ``C * Product(kernel_size)`` is the number of elements within a block (each block contains ``Product(kernel_size)`` vectors containing values from each channel ``C``),
* ``L`` is the total number of blocks calculated as follows:

.. math::

>     L = \prod_{d=1}^{2} \lfloor \frac{({\text{{output\_size}}[d] + \text{{pads\_begin}}[d] + \text{{pads\_end}}[d] - \text{{dilation}}[d] \times (\text{{kernel\_size}}[d] - 1) - 1}}{{\text{{stride}}[d]}}) + 1\rfloor

Reviewer 3: Small issue in built docs, backslash is not formatting how expected: 
![image](https://github.com/openvinotoolkit/openvino/assets/52485122/bccb8d80-16bb-4844-af2f-3d5e58a1d5bb)

Author: It's weird, since it's valid LaTeX syntax. I also tried using `\textunderscore`, but it doesn't render correctly either, and `output_size` is rendered with `size` as a subscript.

In the end I replaced the equation with its text representation, since I couldn't find an example of a variable name with `_` in other docs.
